Feature/files upload

samshara opened the pull request:
- fixes #1854 
- fixes #1855

## Changes

* Add files upload in parallel
* Support Google Drive file upload
* Support Dropbox file upload

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [ ] translations

samshara committed:
Add files upload feature in bulk upload

samshara committed:
Add parallel files upload using Google Drive and Dropbox

tnagorra commented on src/utils/common.js:
```diff
@@ -246,3 +246,14 @@ export function hasKey(data, key, keySelector, childrenSelector) {
     return false;
 }
+export function formatBytes(bytes, decimals = 2) {
+    if (bytes === 0) return '0 Bytes';
```
```suggestion
    if (bytes === 0) {
        return '0 B';
    }
```

tnagorra commented on src/utils/common.js:
```diff
@@ -246,3 +246,14 @@ export function hasKey(data, key, keySelector, childrenSelector) {
+    if (bytes === 0) return '0 Bytes';
+
+    const k = 1024;
+    const dm = decimals < 0 ? 0 : decimals;
```
```suggestion
    const dm = Math.min(0, decimals);
```

tnagorra commented on src/utils/common.js:
```diff
@@ -246,3 +246,14 @@ export function hasKey(data, key, keySelector, childrenSelector) {
+
+    const i = Math.floor(Math.log(bytes) / Math.log(k));
+
+    return `${parseFloat((bytes / (k ** i)).toFixed(dm))} ${sizes[i]}`;
```
We should not return formatted string.
Instead lets return an object with value and suffix so that Numeral can show it later on.

tnagorra replied:
Or, even better let's just create a component to render this.
Ref: https://github.com/idmc-labs/helix-client/blob/develop/src/components/Downloads/DownloadedItem/index.tsx

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/types.ts:
```diff
@@ -0,0 +1,28 @@
+
+export type DropboxFile = Omit<FileLike, 'file'> & {
+    link: string;
+}
```
```suggestion
export type GoogleFileType = Omit<FileLike, 'file'> & {
    mimeType: string;
}

export type DropboxFileType = Omit<FileLike, 'file'> & {
    link: string;
}
```

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/UploadedList/UploadedItem/index.tsx:
```diff
@@ -0,0 +1,14 @@
+
+function UploadedItem(props: Props) {
+    return (
+        <div>{props.name}{props.isUploaded ? 'uploaded' : 'processing'}</div>
```
Shouldn't they be in multiple lines?
Isn't there any linter error?

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/UploadItem/index.tsx:
```diff
@@ -0,0 +1,51 @@
+        }
+    }, [active, uploadFile, data]);
+
+    return <></>;
```
```suggestion
    return null;
```

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/UploadItem/index.tsx:
```diff
@@ -0,0 +1,51 @@
+        if (active) {
+            uploadFile(data);
+        }
+    }, [active, uploadFile, data]);
```
I think we can use useRequest with skip instead of useLazyRequest.

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/GoogleDriveFilesUpload/index.tsx:
```diff
@@ -0,0 +1,121 @@
+                    keySelector={fileKeySelector}
+                    rendererParams={fileRendererParams}
+                />
+            )}
```
I don't think we need this check for files and files length

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/DropboxFilesUpload/index.tsx:
```diff
@@ -0,0 +1,106 @@
+            link: v.link,
+            isUploaded: false,
+        }));
+        setFiles((oldFiles: DropboxFile[]) => ({
+            ...oldFiles,
+            ...values,
+        }));
+        onChange(values);
```
Why do we need to call onChange?

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/FilesUpload/index.tsx:
```diff
@@ -0,0 +1,118 @@
+            >
+                Browse Files
+            </FileInput>
+            {files && files.length > 0 && (
+                <List
+                    data={files}
+                    renderer={UploadItem}
+                    keySelector={fileKeySelector}
+                    rendererParams={fileRendererParams}
+                />
+            )}
```
The file upload mechanism should be global IMO. (shared for drive, dropbox and file uploads)

What do you guys think @AdityaKhatri @frozenhelium

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/FilesUpload/index.tsx:
```diff
@@ -0,0 +1,118 @@
+        <div className={_cs(styles.filesUpload, className)}>
+            <FileInput
+                className={styles.fileInput}
+                value={fileList}
```
Do we need to pass value to FileInput if we immediately upload these files?

samshara committed:
Collect all files in a queue and batch process them to upload

** Remove the success/failure files from the queue
** Add failed files to another array where they can be triggered individually

samshara committed:
Show uploading files in list and failed files in list

tnagorra commented on src/utils/common.js:
```diff
@@ -246,3 +246,17 @@ export function hasKey(data, key, keySelector, childrenSelector) {
+    const i = Math.floor(Math.log(bytes) / Math.log(k));
+
+    return {
+        value: parseFloat((bytes / (k ** i)).toFixed(dm)),
```
We do not do parseFloat here.
We let that decision for Numeral (which can handle dynamic precision as well)

tnagorra commented on src/utils/common.js:
```diff
@@ -246,3 +246,17 @@ export function hasKey(data, key, keySelector, childrenSelector) {
+
+    const k = 1024;
+    const dm = Math.min(0, decimals);
+    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
```
```suggestion
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
```

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/DropboxFilesUpload/index.tsx:
```diff
@@ -0,0 +1,53 @@
+
+interface Props {
+    className?: string;
+    onChange: (v: FileLike[]) => void;
```
Shouldn't this be onAdd or onUpload?

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/FilesUpload/index.tsx:
```diff
@@ -0,0 +1,58 @@
+
+interface Props {
+    className?: string;
+    onChange: (v: FileLike[]) => void;
```
Shouldn't this be onAdd or onUpload?

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/GoogleDriveFilesUpload/index.tsx:
```diff
@@ -0,0 +1,71 @@
+
+interface Props {
+    className?: string;
+    onChange: (v: FileLike[]) => void;
```
Shouldn't this be onUpload or onAdd?

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/GoogleDriveFilesUpload/index.tsx:
```diff
@@ -0,0 +1,71 @@
+
+    const setAccessToken = (token: string) => {
+        accessToken.current = token;
+    };
```
Let's use useCallback

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/GoogleDriveFilesUpload/index.tsx:
```diff
@@ -0,0 +1,71 @@
+        onChange,
+    } = props;
+
+    const accessToken = useRef<string>();
```
```suggestion
    const accessToken = useRef<string | undefined>();
```

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/Upload/UploadItem/index.tsx:
```diff
@@ -0,0 +1,126 @@
+        },
+    });
+
+    const handleRetriggerClick = useCallback(() => {
+        setIsTriggeredEnable(true);
+        retrigger();
+    }, [retrigger]);
```
We can re-trigger? How does this work?
Simply moving the file from failedFiles to files?

samshara committed:
Create Sources uploaded view and trigger source creation

samshara committed:
Add search functionality in sources uploaded pane

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/FilesUploaded/FilesSearchInput/index.tsx:
```diff
@@ -0,0 +1,58 @@
+            onSearchFiles(filteredOptions);
+        } else {
+            onSearchFiles(undefined);
+        }
```
Why pass  filtered files outside?
This should be handled during render

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/FilesUploaded/index.tsx:
```diff
@@ -0,0 +1,80 @@
+        onDeleteFile,
+    }), [onDeleteFile]);
+
+    const handleSearchFiles = useCallback((matchingFiles: number[] | undefined) => {
+        setSearchResults(matchingFiles);
+    }, []);
```
Do we need to create a function? Why not just alias

const handleSearchFiles = setSearchResults;

tnagorra commented on src/newViews/Tagging/Sources/BulkUpload/FilesUploaded/index.tsx:
```diff
@@ -0,0 +1,80 @@
+                contentClassName={styles.files}
+                sub
+            >
+                <FilesSearchInput
+                    className={styles.search}
+                    options={files}
+                    onSearchFiles={handleSearchFiles}
+                />
```
The search state should be moved to this parent component and parent should implement the filtering logic

tnagorra committed:
Remove bulkupload page
